TDDWithURMAndKotlin/part2_increment.md

```diff
@@ -0,0 +1,157 @@
+
+```kotlin
+fun increment(registry: Registry, position: Int){
+    require(position > 0) { "Position must be positive number" }
+    checkNotNull(registry.getValueAtPosition(position)) { "Register must be initialized" }
+}
+```
+
+Now our tests are passing again. Let's write the actual implementation.
+
+## The real implementation
+
+The test for the real implementation will need a litle more setup. We know that the current approach will throw an expection if our test is just:
+
+```kotlin
+@ParameterizedTest
+@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
+fun `increment function adds 1 to value of position`(position: Int) {
+    increment(registry, position)
+
+    assertEquals(1, registry.getValueAtPosition(position))
+}
+```
+
+This is because we wrote a test that will throw an exception when a registry does not contain a value and someone calls the `increment` function on it. But fortunately we already wrote tests for the `zero` function which means we can initialize our registry before calling increment:
+
+```kotlin
+@ParameterizedTest
+@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
+fun `increment function adds 1 to value of position`(position: Int) {
+    zero(registry, position)
+    increment(registry, position)
+
+    assertEquals(1, registry.getValueAtPosition(position))
+}
+```
+
+To fix the breaking test we need to modify our function:
+
+```kotlin
+fun increment(registry: Registry, position: Int) {
+    require(position > 0) { "Position must be positive number" }
+
+    val currentValue = registry.getValueAtPosition(position)
+    checkNotNull(currentValue) { "Register must be initialized" }
+
+    registry.setValueAtPosition(position, currentValue + 1)
+}
+```
+
+We store the value of `registry.getValueAtPosition(position)` and then call the contract `checkNotNull` (I'll write about contracts soon, I promise!)
+
+But testing with a single `1` is boring… Let's write something that operates over different values. To test dinamycally we will need to check a parameter, then call `increment` and get an expected value. We can simply call `assertEquals(parameter + 1, registry.getValueAtPosition(parameter))` but somehow it feels like cheating…
+
+Using _parameterized tests_ from JUnit 5 we can create a single test that will cover different cases in a single method, and we will use `@CsvSource` to pass different parameters. `@CsvSource` takes a group of comma-separated-strings for every set of parameters and our method will ask for this parameters in the appropiate types:
+
+```kotlin
+@ParameterizedTest
+@CsvSource("50,1,2", "50,5,6", "50,100,101")
+fun `increment function adds 1 to any integer value`(
+    position: Int,
+    original: Int,
+    expected: Int
+){
+    registry.setValueAtPosition(position, original)
+
+    increment(registry, position)
+
+    assertEquals(expected, registry.getValueAtPosition(position))
+}
+```
+
+This allow the test to be more dynamic and we can add more cases having more flexibility!
```

Comment:
This allowS **(capital S)**

Reply:
what? CAPITAL S LIKE IN ALL CAPS BUT JUST FOR THE **S**?